### ISSUE TITLE
Can place traps and doors before you have a workshop

### ISSUE THREAD

**Issue by Loobinex** [OpenDungeons/OpenDungeons#1207]
In version git20160601 you can buy and place traps/doors before you even have the workshop researched, let alone built. This is a bit silly, and also confusing for the player. I for example accidentally spent some gold on a canon-trap before I realized my workshop was not yet finished researching, this could be a lot worse for new players that don't know they need a workshop to build traps.

My suggestion is to disable all trap/door buttons until you build a workshop.

**Comment by hwoarangmy** [comment]
Here, I would be for displaying a message (and a voice alert) to say that you need a workshop when you place a trap.
That's because if we don't enable traps, what happens if you own a workshop but no creature to work there ? Should we disable the traps ? What happens if your workshop gets destroyed ?

IMHO, the simpliest solution is the best and a message is simpler ^^

**Comment by Loobinex** [comment]
A message would be an improvement over the current scenario, but there is a downside.

As a general rule for any interface you don't want a user to have to take an action first (click a button, place a trap, whatever) to see what happens. You want to be clear in advance what's the consequence.
Not enabling traps before you have a workshop provides a much more intuitive interface.

Trap-buttons should be enabled when you have a workshop but no workers - and this is a perfect situation for a voice alert.
Disabling the traps again when your workshop is destroyed fine.

You still have a very simple solution like this by the way, one that comes down to nothing more and nothing less then 'Only have the trap buttons enabled when/while the player owns a workshop'.

**Comment by akien-mga** [comment]
But even if you don't have a workshop, or your workshop was destroyed, you might still have workers carrying crafted traps around (either from your previous workshop, or from an enemy creature that dropped it, or some other bonus that was carried by a rogue creature), so you should still be able to place traps if you spent time researching them.

IMO adding a message + ensuring that you have to research the workshop before researching traps would be enough.

**Comment by Loobinex** [comment]
I guess that means it's decided.

So how does it currently work if you find a trap you have not researched yourself? You should be able to place that too right?
If so, you could also always enable placing traps(not buying them) that you currently own. That would work in both situations.

**Comment by akien-mga** [comment]
It's not decided, we're discussing :)

The main problem with your idea IMO is that currently, the players have no easy way to see if they have a workshop, or have crafted traps that they can use, etc. So enabling/disabling buttons based on such information that is not evident to the player is IMO more confusing than letting the player place a trap to get a warning that they don't have/no longer have a workshop.

If the players were forced to research the workshop before traps, they'll maybe make the mistake on their first game, see the warning, and then understand that traps need a workshop to be crafted. Disabling the buttons (with an explanatory tooltip) might be better for such first-time players, but on the long run I think it doesn't make things easier to toggle buttons on/off based on information the player doesn't have.

Of course we could also try to make this kind of information more visible :)

**Comment by Loobinex** [comment]
This is a way to help make it visible for the player he has no workshop. I don't know how you do your research, but I have a preferred path and when I'm waiting for the level to get properly started when my first imp digs my first rooms I just select everything to be researched and be done with it.

Even for more experienced players it is easy to forget if you already have build a workshop or not. With all your traps being able to be built you get the suggestion you already have. If you can't build traps it is immediately clear you have yet to build your workshop.

EDIT: It was a serious question by the way, can you place traps you haven't yet researched if you capture them from your rival?

**Comment by hwoarangmy** [comment]
> It was a serious question by the way, can you place traps you haven't yet researched if you capture them from your rival?

No

**Comment by Loobinex** [comment]
I think you should be able to.

**Comment by hwoarangmy** [comment]
Well, there is a difference between a crafted trap (workshop object) and the trap itself (research). And for the logic behind this, I would say that it sounds pretty normal to me that you don't know how to use something until you researched it. If you find some alien weapon, you won't know what it is for and how to use it. You will need to investigate and find out this information. Same here.

**Comment by akien-mga** [comment]
We could however change the way we think about traps and their research:
- The trap buttons would all be shown by default, but deactivated
- Researching a trap would provide the ability to _craft_ the trap
- All traps can be placed if you have them in your inventory, i.e. if you steal a trap from the enemy, you can place it, even if you don't know how to craft it (further on the alien weapon example, if it has a trigger that you can pull, you _might_ understand how to use it even if you don't know how to craft it ;))
- The trap buttons would show the number of crafted traps you have in your inventory

So basically the trap buttons would be changed from a "place a trap and order its construction" to a "place a trap from the inventory" behaviour. Then the question is how to _order_ the construction of traps so that you have some to place. Maybe by adding small +/- buttons on top of the trap icon button could be use to handle a build queue, etc.

This is just a wild idea, it needs to be further specified as we would have to change a bit the way the trap logic works. WDYT about the overall concept?

**Comment by Loobinex** [comment]
I think you indeed research the ability to create a trap, not to use it. Not only from a conceptual point of view, but from a gameplay perspective this is also the superior option. For one it allows what I said above, to place traps you steal from your enemy even if you don't have a workshop/haven't researched traps yet, but also importantly it allows the creation of maps where trap boxes are provided by the mapmaker where the workshop isn't provided or is not available yet.

This fits your concept akien-mga.
